File: solutions/fortran/isbn-verifier/1/isbn_verifier.f90
module isbn_verifier
  implicit none

contains

  function isValid(isbn) result(valid)
    character(*), intent(in) :: isbn
    logical :: valid
    integer :: i, j, digit, sums
    valid= .true.
    !check for invalid lengths
    if (len(isbn)/=10 .and. len(isbn)/=13) then
      valid=.false.
    !check for valid characters on a 10 digit string
    else if (scan(isbn,"ABCDEFGHIJKLMNOPQRSTUVWYZ-")>0 .and. len(isbn)==10) then
      valid=.false.
    !check for valid characters on a 13 digit string
    else if (scan(isbn,"ABCDEFGHIJKLMNOPQRSTUVWYZ")>0 .and. len(isbn)==13) then
      valid=.false.
    !check for invalid X in a different position
    else if (scan(isbn,"X")>0 .and. isbn(len(isbn):len(isbn))/="X") then
      valid=.false.
    else 
      !calculate checksum
      sums=0
      j=10
      do i=1,len(isbn)
        if (isbn(i:i)/="-") then
          if (isbn(i:i)=="X") then
            digit=10
          else
            read(isbn(i:i),'(i1)') digit
          end if
          sums=sums+digit*j 
          j=j-1
        end if
      end do
      valid = mod(sums,11)==0
    end if
  end function isValid

end module isbn_verifier

File: solutions/python/isbn-verifier/1/isbn_verifier.py
digit="0123456789"

def is_valid(isbn):
    valid=True
    isbn=isbn.replace("-","")
    if not len(isbn) == 10:
        valid = False
    elif isbn[:9].isnumeric() and (isbn[9:]=='X' or isbn[9:].isnumeric()):
        checksum=0
        for i in range(0,9):
            checksum=checksum+int(isbn[i])*(10-i)
        if isbn[9:]=='X':
            checksum=checksum+10
        else:
            checksum=checksum+int(isbn[9:])
        if not checksum % 11 == 0:
            valid = False
    else:
        valid = False
    return valid
    pass
